apps/emqx/src/emqx_trace/emqx_trace.erl

```diff
@@ -688,22 +690,33 @@ stop_trace(#{id := HandlerID, dst := Basename, filter := {Type, Filter}} = Handl
 clean_stale_trace_files() ->
     TraceDir = trace_dir(),
     case file:list_dir(TraceDir) of
-        {ok, AllFiles} when AllFiles =/= ["zip"] ->
-            KeepFiles = [
+        {ok, []} ->
+            ok;
+        {ok, ["zip"]} ->
+            ok;
+        {ok, Files} ->
+            %% NOTE: Cleaning both stale and foreign files.
+            TraceFiles = [
                 log_filename(Name, StartAt)
              || #?TRACE{name = Name, start_at = StartAt} <- list_traces()
             ],
-            case AllFiles -- ["zip" | KeepFiles] of
-                [] ->
-                    ok;
-                DeleteFiles ->
-                    DelFun = fun(F) -> file:delete(filename:join(TraceDir, F)) end,
-                    lists:foreach(DelFun, DeleteFiles)
-            end;
-        _ ->
+            StaleFiles = [Filename || Filename <- Files, trace_file_is_stale(Filename, TraceFiles)],
+            lists:foreach(
+                fun(Filename) ->
+                    ?tp(debug, "trace_cleaning_stale_file", #{filename => Filename}),
+                    file:delete(filename:join(TraceDir, Filename))
+                end,
+                StaleFiles
+            );
+        _Error ->
             ok
     end.
 
+trace_file_is_stale("zip", _TraceFiles) ->
+    false;
+trace_file_is_stale(Filename, TraceFiles) ->
+    lists:all(fun(Basename) -> string:prefix(Filename, Basename) =:= nomatch end, TraceFiles).
```

Comment:
maybe add a is_active_trace/1 which parses the filename and lookup in the table ?

Reply:
That was actually the approach I tried first. It became quite hairy, mostly because of non-trivial regexp-based parsing and correct matching parsed dates against `started_at`. Of course O(N²) is not great, what do you think about [7b66600](https://github.com/emqx/emqx/pull/15986/commits/7b66600447d79ae37ddbb47344717df74c265368)?